torch/_functorch/autograd_function.py

```diff
@@ -0,0 +1,196 @@
+        # (because autograd.Function happens before the Python dispatch key)
+        # and only traces the forward pass.
+        if torch._C._are_functorch_transforms_active():
+            return super().__call__(*args, **kwargs)
+        autograd_function = args[0]
+        return autograd_function.apply(*args[1:], **kwargs)
+
+
+# "custom_function_call"
+# This is the mechanism for an autograd.Function that works with functorch transforms.
+# It wraps an autograd.Function; interactions with functorch transforms are defined
+# via PyDispatcher and PyOperator rather than through the traditional PyTorch
+# dispatcher.
+custom_function_call = CustomFunctionPyOperator()
+
+
+# The grad rule for custom_function_call is to construct a new _SingleLevelFunction
+# (autograd.Function that only works with a single layer (level) of functorch) that:
+# - unwraps the inputs
+# - redispatches to custom_function_call
+# - wraps the outputs
+# and whose backward pass calls the original autograd.Function's backward.
+#
+# Why do we need to redispatch to custom_function_call?
+# -----------------------------------------------------
+# This is consistent with how ATen operators work with functorch's grad transform:
+# they always redispatch to the original operator.
+# Consider torch.sin, and let's say we do grad0(grad1(torch.sin))(x)
+#
+# grad1 will:
+# - set up the autograd graph
+# - unwrap the inputs
+# - redispatch to at::sin (*)
+# - rewrap the outputs on the return
+#
+# On the redispatch in (*), grad0 will:
+# - set up the autograd graph
+# - unwrap the inputs
+# - redispatch to at::sin
+# - rewrap the outputs on the return
+#
+# To "set up the autograd graph", we generate a _SingleLevelFunction
+# and apply it.
+@custom_function_call.py_impl(TransformType.Grad)
+def custom_function_call_grad(interpreter, autograd_function, *operands):
+    maybe_interpreter = interpreter
+    level = maybe_interpreter.level()
+
+    # TODO: The name of the grad_fn is GeneratedBackward. This isn't a great UX,
+    # but in theory functorch users shouldn't be peeking at the grad_fn.
+    # We should try to generate a better name for this.
+    class Generated(_SingleLevelFunction):
+        @staticmethod
+        def forward(*operands):
+            unwrapped_operands = pytree.tree_map_only(
+                torch.Tensor,
+                lambda x: _unwrap_for_grad(x, level),
+                operands)
+            with torch.enable_grad(), maybe_interpreter.lower():
+                output = custom_function_call(autograd_function, *unwrapped_operands)
+
+            # autograd.Function's ctx.mark_dirty expect a returned input
+            # to have the same object identity as the input.
+            # Mode-only functorch will greatly simplify this logic.
+            return wrap_outputs_maintaining_identity(
+                output,
+                unwrapped_operands,
+                operands,
+                level)
```

Comment:
@soulitzer -- It looks like autograd.Function only preserves object identity if the input has requires_grad=True. Is this intentional?

The effect of this is that when someone runs e.g. NumpyExp_ on a Tensor that doesn't require grad inside of functorch.grad, it complains that the output being returned is not the same as the input. The output is not the same as the input because this logic highlighted fails since the the output of autograd.Function does not have the same object identity as the input unless the input has requires_grad=True:

```py
import torch
import numpy as np

class NumpyExp_(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        x_np = x.numpy()
        np.exp(x_np, x_np)
        ctx.mark_dirty(x)
        ctx.save_for_backward(x)
        return x

    @staticmethod
    def backward(ctx, grad_output):
        output, = ctx.saved_tensors
        return NumpyMul.apply(grad_output, output)

x = torch.tensor(1., requires_grad=True)
z = x.clone()
y = NumpyExp_.apply(z)
print(y is z)
```

Reply:
Yeah that looks weird, I think it would be good to make that behavior consistent between using requires_grad=True/False. Should be a small fix.